src/datadog/trace_segment.cpp

```diff
@@ -1,40 +1,56 @@
 #include "trace_segment.h"
 
 #include <cassert>
-#include <charconv>
-#include <limits>
 #include <string>
-#include <system_error>
 #include <unordered_map>
 #include <utility>
+#include <vector>
 
 #include "collector.h"
 #include "collector_response.h"
 #include "dict_writer.h"
 #include "error.h"
+#include "hex.h"
 #include "logger.h"
 #include "optional.h"
 #include "span_data.h"
 #include "span_sampler.h"
 #include "tag_propagation.h"
 #include "tags.h"
 #include "trace_sampler.h"
+#include "w3c_propagation.h"
 
 namespace datadog {
 namespace tracing {
 namespace {
 
-template <typename Integer>
-std::string hex(Integer value) {
-  // 4 bits per hex digit char, and then +1 char for possible minus sign
-  char buffer[std::numeric_limits<Integer>::digits / 4 + 1];
-
-  const int base = 16;
-  auto result =
-      std::to_chars(std::begin(buffer), std::end(buffer), value, base);
-  assert(result.ec == std::errc());
+// Encode the specified `trace_tags`. If the encoded value is not longer than
+// the specified `tags_header_max_size`, then set it as the "x-datadog-tags"
+// header using the specified `writer`. If the encoded value is oversized, then
+// write a diagnostic to the specified `logger` and set a propagation error tag
+// on the specified `local_root_tags`.
+void inject_trace_tags(
+    DictWriter& writer,
+    const std::vector<std::pair<std::string, std::string>>& trace_tags,
+    std::size_t tags_header_max_size,
+    std::unordered_map<std::string, std::string>& local_root_tags,
+    Logger& logger) {
+  const std::string encoded_trace_tags = encode_tags(trace_tags);
 
-  return std::string{std::begin(buffer), result.ptr};
+  if (encoded_trace_tags.size() > tags_header_max_size) {
+    std::string message;
+    message +=
+        "Serialized x-datadog-tags header value is too large.  The configured "
+        "maximum size is ";
+    message += std::to_string(tags_header_max_size);
+    message += " bytes, but the encoded value is ";
+    message += std::to_string(encoded_trace_tags.size());
+    message += " bytes.";
+    logger.log_error(message);
+    local_root_tags[tags::internal::propagation_error] = "inject_max_size";
```

Comment:
Why on local-root instead of the span specifically?
What if the same span is propagated more than once?
What if multiple child spans of the local-root experienced the same error?

Reply:
`tags::internal::propagation_error` is specified in the horizontal propagation RFC. There it's defined to be set on the local root span, and implementations are allowed to either keep the first error or to keep the last error. I chose the latter.